### DISCUSSION TITLE
Renovate keeps reverting lockfileversion to 1 despite engine constraint

### DISCUSSION THREAD

**Discussion by ge-vlad-romanov** [renovatebot/renovate#13027] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

29.33

### Please select which platform you are using if self-hosting.

GitHub Enterprise Server

### Describe the bug

Our git repo has a package-lock.json file with lockfileversion 2 as we are now using Node 16.13.0 and npm 8.1.0. Our understanding was that the latest versions of Renovate would respect the lockfileversion in package-lock.json, however it seems to revert to lockfileversion 1 in every PR, causing diffs like this:
<img width="150" alt="Screenshot 2021-12-09 at 07 29 41" src="https://user-images.githubusercontent.com/32774223/145353441-ddafd041-c5ed-4b30-b945-6a90c56ef199.png">

Our config.js file:

```js
module.exports = {
  endpoint: 'https://github.build.ge.com/api/v3',
  gitAuthor: 'Renovate Bot <noreply@github.build.ge.com>',
  platform: 'github',
  logFileLevel: 'debug',
  logLevel: 'debug',
  onboarding: true,
  onboardingConfig: { 'extends': ['config:base'] },
  baseBranches: [<branch name>],
  includeForks: true,
  repositories: [<repository_name>],
};
```

I have tried adding an npm.engine constraint to the package.json file, which Renovate picks up according to the debug logs but it has no effect that I can tell (although I am not clear if it's meant to work if the binarySource is global).

Any advice on what we can do to stop it reverting the lockfileversion would be greatly appreciated - as it stands we have had to stop using Renovate as every PR needed a manual intervention to re-generate the lock file with the correct version.

P.S. in the logs section below I have attached the parts that I think are most relevant (the constraint being logged and the logs of it generating the package-lock file). There are a lot of logs though so let me know if there is any other part that would be of interest. I do see this line which seems relevant, but I don't really know what "do my best" means!
```
"stderr": "npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!\n"
```

### Relevant debug logs

<details><summary>Logs</summary>

```
{
                 "depType": "engines",
                 "depName": "npm",
                 "currentValue": "^8.1.0",
                 "datasource": "npm",
                 "commitMessageTopic": "npm",
                 "prettyDepType": "engine",
                 "lockedVersion": null,
                 "depIndex": 143,
                 "updates": [],
                 "warnings": [],
                 "versioning": "npm",
                 "sourceUrl": "https://github.com/npm/cli",
                 "homepage": "https://docs.npmjs.com/",
                 "currentVersion": "8.2.0"
               }
             ],
             "packageJsonName": <repo name>,
             "packageFileVersion": "0.0.0",
             "packageJsonType": "app",
             "npmLock": "package-lock.json",
             "managerData": {"yarnZeroInstall": false},
             "skipInstalls": true,
             "constraints": {"npm": "^8.1.0"},
             "lockFiles": ["package-lock.json"]
           }
           
           ....

DEBUG: Getting updated lock files (repository=<repo name>, branch=renovate/code-analysis-packages)
DEBUG: Writing package.json files (repository=<repo name>, branch=renovate/code-analysis-packages)
       "packageFiles": ["package.json"]
DEBUG: Writing package-lock.json (repository=<repo name>, branch=renovate/code-analysis-packages)
DEBUG: Massaging npm lock file before writing to disk (repository=<repo name>, branch=renovate/code-analysis-packages)
DEBUG: Writing any updated package files (repository=<repo name>, branch=renovate/code-analysis-packages)
DEBUG: Writing package.json (repository=<repo name> branch=renovate/code-analysis-packages)
DEBUG: No npmrc file found in repository (repository=<repo name>, branch=renovate/code-analysis-packages)
DEBUG: Writing updated .npmrc file to /tmp/renovate/repos/github/<repo name>/.npmrc (repository=<repo name> branch=renovate/code-analysis-packages)
DEBUG: Generating package-lock.json for . (repository=<repo name>, branch=renovate/code-analysis-packages)
DEBUG: Spawning npm install to create /tmp/renovate/repos/github/<repo name>/package-lock.json (repository=<repo name> branch=renovate/code-analysis-packages)
DEBUG: Updating lock file only (repository=<repo name>, branch=renovate/code-analysis-packages)
DEBUG: No node constraint found - using latest (repository=<repo name>, branch=renovate/code-analysis-packages)
DEBUG: Executing command (repository=<repo name>, branch=renovate/code-analysis-packages)
       "command": ["npm install --package-lock-only --no-audit --ignore-scripts"]
DEBUG: exec completed (repository=<repo name>, branch=renovate/code-analysis-packages)
       "cmd": "npm install --package-lock-only --no-audit --ignore-scripts",
       "durationMs": 18449,
       "stdout": "added 3454 packages in 15.832s\n",
       "stderr": "npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!\n"
DEBUG: package-lock.json needs updating (repository=<repo name>, branch=renovate/code-analysis-packages)
DEBUG: Updated 1 lock files (repository=<repo name>, branch=renovate/code-analysis-packages)
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment, marked as answer]
You need to use the `binarySource=docker` and the `renovate/renovate:slim` image to allow renovate to respect contraints until https://github.com/renovatebot/renovate/issues/11984 is done.

**Reply by ge-vlad-romanov** [reply]
Thank you for the quick reply! If not using binarySource=docker, what is the current logic of how Renovate chooses which Node/npm version to use? From the debug logs I gathered that it uses the 'latest' node, which I assume is 16, and therefore should use npm 7 or 8, so I don't understand why it's reverting the lockfileversion. But I am almost definitely lacking understanding here! 

Regarding https://github.com/renovatebot/renovate/issues/11984 - when this is implemented, will anything be needed to be configured besides setting binarySource=install in the config.js file to get it to respect the constraints?

**Reply by rarkins** [reply]
> Thank you for the quick reply! If not using binarySource=docker, what is the current logic of how Renovate chooses which Node/npm version to use? From the debug logs I gathered that it uses the 'latest' node, which I assume is 16, and therefore should use npm 7 or 8, so I don't understand why it's reverting the lockfileversion. But I am almost definitely lacking understanding here!

Your misunderstanding is because Renovate doesn't "choose which" npm to use at all, when binarySource=global. Choosing would mean installing, and it does not runtime install today. It just calls whatever `npm` is in the `$PATH` (i.e. "globally" installed, hence global). If you want a different default version then you'd need to customize the container you're running.

> Regarding #11984 - when this is implemented, will anything be needed to be configured besides setting binarySource=install in the config.js file to get it to respect the constraints?

Nothing additional should be needed. Once we consider the feature to be well tested, we'll probably default to binarySource=install in our distributed images so that you don't need to configure even that.

**Reply by ge-vlad-romanov** [reply]
Thank you @rarkins - I think I read too much into the debug logs. Currently we are running it off a VM, so I assume I will need to ensure that the default Node and NPM versions on that machine are 16 and 8 respectively (if that's what I want to use)?

I will keep an eye on the binarySource=install addition too, thanks for letting me know.

**Reply by ge-vlad-romanov** [reply]
I have updated the Node and npm versions on the VM via nvm to the expected values:
node --version -> 16.13.0
npm --version -> 8.1.0

and verified that both appear in the $PATH, e.g. <some_path>/.nvm/versions/node/v16.13.0/bin and <some_path>/.nvm/versions/node/v16.13.0/bin/npm

but it still reverts the lockfileversion to 1 and has the same line about trying to 'do its best'. Is there anything else I need to add, e.g. to the docker run command to pass the environment through? E.g. right now it runs 

```
docker run --rm -e LOG_LEVEL=debug -e HTTPS_PROXY -e HTTP_PROXY -e RENOVATE_TOKEN -e GITHUB_COM_TOKEN -v "/config.js:/usr/src/app/config.js"  renovate/renovate:29.33
```

If there is nothing left for me to try then I might just wait for the updated distributed images.

**Reply by rarkins** [reply]
It looks like you are running renovate inside a container and not directly on the VM? If so then the versions on the VM don't help

**Reply by ge-vlad-romanov** [reply]
My bad, I am going round in circles here, apologies. The script to run renovate is on a VM, but the script runs the above `docker run` command which runs the Renovate image. Can I customise the versions of node and npm that run inside the docker image via the docker run command? I see now that it installs node v14.18.2 in one of its image layers, which is what I should have looked at initially 🤦

**Reply by viceice** [reply]
if you run renovate with docker anyways, use our slim image, it' smuch more smaller and uses docker sidecar containers to met constraints

checkout https://github.com/renovatebot/docker-renovate#samples

you only need to pass `/tmp` and docker socket volumes

**Reply by ge-vlad-romanov** [reply]
Thanks, I will try it out! I also saw there was a renovate npm package that can be used as an alternative, does that help in my situation at all or does it behave the same way as the docker images?

**Reply by rarkins** [reply]
If you install via npm - which you can do directly into your VM - then you need to make sure that all package managers you may need are already installed (e.g. not just npm but perhaps yarn, bundler, poetry, composer, etc).

**Reply by viceice** [reply]
if you have docker available use the docker image, so you don't need to install any dependencies

**Reply by ge-vlad-romanov** [reply]
Thank you both - I have gotten much further with the slim image and it is looking promising. Is there anything additional I need to do to pass proxy values through to the sidecar containers it spins up? I have -e HTTPS_PROXY -e HTTP_PROXY in the `docker run` command, which worked for the full image, but in my slim image it runs into proxy errors when talking to the npm registry, e.g.
`npm ERR! network request to https://registry.npmjs.org/npm failed, reason: getaddrinfo ENOTFOUND`

**Reply by rarkins** [reply]
I would have expected it to work. Please start a separate discussion though

**Reply by viceice** [reply]
as workaround you can set custom env values to `config.js`
https://docs.renovatebot.com/self-hosted-configuration/#customenvvariables

**Reply by ge-vlad-romanov** [reply]
Thank you for all your quick help today, and I apologise for piling on the questions. If anything else comes up I will raise it separately - have a good rest of the week 👍

**Reply by xoxys** [reply]
Is there really no way to get it working correctly with the "full" image? Is it intended that the full image always downgrades the package lock file version to v1?

**Reply by rarkins** [reply]
Set binarySource=install, read the docs for more details
